File: Follower/service/follower_service.go
package service

import (
	"context"
	"follower-service/model"
	"follower-service/repository"
)

type FollowerService struct {
	Repo *repository.FollowerRepository
}

func NewFollowerService() *FollowerService {
	return &FollowerService{
		Repo: &repository.FollowerRepository{},
	}
}

func (s *FollowerService) CreateUser(ctx context.Context, userID int, name, username string) error {
	return s.Repo.CreateUser(ctx, userID, name, username)
}

func (s *FollowerService) FollowUser(ctx context.Context, followerID, followingID int) error {
	// Check if already following
	isFollowing, err := s.Repo.IsFollowing(ctx, followerID, followingID)
	if err != nil {
		return err
	}

	if isFollowing {
		return nil // Already following, no need to create duplicate relationship
	}

	return s.Repo.FollowUser(ctx, followerID, followingID)
}

func (s *FollowerService) UnfollowUser(ctx context.Context, followerID, followingID int) error {
	return s.Repo.UnfollowUser(ctx, followerID, followingID)
}

func (s *FollowerService) IsFollowing(ctx context.Context, followerID, followingID int) (bool, error) {
	return s.Repo.IsFollowing(ctx, followerID, followingID)
}

func (s *FollowerService) GetFollowers(ctx context.Context, userID int) ([]model.User, error) {
	return s.Repo.GetFollowers(ctx, userID)
}

func (s *FollowerService) GetFollowing(ctx context.Context, userID int) ([]model.User, error) {
	return s.Repo.GetFollowing(ctx, userID)
}

func (s *FollowerService) GetUserStats(ctx context.Context, userID int) (*model.UserFollowStats, error) {
	return s.Repo.GetUserStats(ctx, userID)
}

// CanUserComment checks if a user can comment on another user's blog using userIds
func (s *FollowerService) CanUserComment(ctx context.Context, commenterID, blogAuthorID int) (bool, error) {
	// Users can always comment on their own blogs
	if commenterID == blogAuthorID {
		return true, nil
	}

	// Check if commenter is following the blog author
	return s.Repo.IsFollowing(ctx, commenterID, blogAuthorID)
}

// GetFollowRecommendations returns recommendations for users to follow
func (s *FollowerService) GetFollowRecommendations(ctx context.Context, userID int, limit int) (*model.RecommendationsResponse, error) {
	if limit <= 0 {
		limit = 10 // default limit
	}
	if limit > 50 {
		limit = 50 // max limit
	}

	recommendations, err := s.Repo.GetFollowRecommendations(ctx, userID, limit)
	if err != nil {
		return nil, err
	}

	return &model.RecommendationsResponse{
		Recommendations: recommendations,
		TotalCount:      len(recommendations),
	}, nil
}

File: Front/TouristApp/src/app/followers/follower.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface FollowRequest {
  followerId: number;
  followingId: number;
}

export interface FollowResponse {
  success: boolean;
  message: string;
  following?: boolean;
}

export interface User {
  userId: number;
  username?: string;
  name?: string;
}

export interface UserFollowStats {
  userId: number;
  followersCount: number;
  followingCount: number;
}

export interface FollowRecommendation {
  recommendedUser: User;
  mutualFollowers: string[];
  recommendationReason: string;
}

export interface RecommendationsResponse {
  recommendations: FollowRecommendation[];
  totalCount: number;
}

@Injectable({
  providedIn: 'root'
})
export class FollowerService {
  private apiUrl = 'http://localhost:8070/followers';

  constructor(private http: HttpClient) { }

  followUser(followerId: number, followingId: number): Observable<FollowResponse> {
    const request: FollowRequest = { followerId, followingId };
    return this.http.post<FollowResponse>(`${this.apiUrl}/follow`, request);
  }

  unfollowUser(followerId: number, followingId: number): Observable<FollowResponse> {
    const request: FollowRequest = { followerId, followingId };
    return this.http.post<FollowResponse>(`${this.apiUrl}/unfollow`, request);
  }

  isFollowing(followerId: number, followingId: number): Observable<{isFollowing: boolean}> {
    return this.http.get<{isFollowing: boolean}>(`${this.apiUrl}/is-following/${followerId}/${followingId}`);
  }

  getFollowers(userId: number): Observable<User[]> {
    return this.http.get<User[]>(`${this.apiUrl}/users/${userId}/followers`);
  }

  getFollowing(userId: number): Observable<User[]> {
    return this.http.get<User[]>(`${this.apiUrl}/users/${userId}/following`);
  }

  getUserStats(userId: number): Observable<UserFollowStats> {
    return this.http.get<UserFollowStats>(`${this.apiUrl}/users/${userId}/stats`);
  }

  canUserComment(commenterId: number, authorId: number): Observable<{canComment: boolean}> {
    return this.http.get<{canComment: boolean}>(`${this.apiUrl}/can-comment/${commenterId}/${authorId}`);
  }

  createUser(userId: number, name: string, username?: string): Observable<FollowResponse> {
    return this.http.post<FollowResponse>(`${this.apiUrl}/users`, { userId, name, username });
  }

  getFollowRecommendations(userId: number, limit: number = 10): Observable<RecommendationsResponse> {
    return this.http.get<RecommendationsResponse>(`${this.apiUrl}/users/${userId}/recommendations?limit=${limit}`);
  }
}